### ISSUE TITLE
LoRA + DeBERTa: loading model gives erratic, non-deterministic results

### ISSUE THREAD

**Issue by jchook** [huggingface/peft#2171]
### System Info
A100 Colab

### Issue

Here is a Colab repro to demonstrate my issue: https://colab.research.google.com/drive/1Z5FL1QDePY0j-8XL0o1V9o27_MbU-02E?usp=sharing

It's very possible that I am doing something fundamentally wrong or using the `peft` framework incorrectly. However, I tried to copy most of this code from [the official LoRA sequence classification example](https://github.com/huggingface/peft/blob/main/examples/sequence_classification/LoRA.ipynb)

On this toy **IMDB sentiment classification example using LoRA + DeBERTA base**, I get consistent ~100% test accuracy when I do any of the following:
- Measure accuracy on the train set during training
- Measure accuracy on the test set immediately after training an epoch
- FFT the base model without the PEFT framework

However, **I get erratic results when saving (per-epoch) and loading a trained LoRA adapter from disk**.

This occurs without any modifications to the code, in the same python session, or across fresh python sessions. So, I have ruled-out residual python state in RAM, etc.

The results appear to shift randomly between the following outcomes: 
- very high accuracy, close to 100% 
- unreasonably low accuracy, close to 0% (i.e. inverted labels)
- something in between

```
{'accuracy': 0.3}
{'accuracy': 0.0}
{'accuracy': 0.979}
{'accuracy': 1.0}
{'accuracy': 0.303}
{'accuracy': 0.812}
{'accuracy': 0.591}
{'accuracy': 1.0}
```

### Who can help?

@BenjaminBossan

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

Here is a Colab repro to demonstrate my issue: https://colab.research.google.com/drive/1Z5FL1QDePY0j-8XL0o1V9o27_MbU-02E?usp=sharing

In the Colab, after running the other cells to install, load the `imdb` dataset, and train a LoRA classifier, I am running this test cell repeatedly and seeing erratic results:

```python
import torch
from peft import PeftModel, PeftConfig
from transformers import AutoModelForCausalLM, AutoTokenizer

peft_model_id = "./output/lora_epoch_1"
config = PeftConfig.from_pretrained(peft_model_id)
inference_model = AutoModelForSequenceClassification.from_pretrained(config.base_model_name_or_path)
tokenizer = AutoTokenizer.from_pretrained(config.base_model_name_or_path)

# Load the Lora model
inference_model = PeftModel.from_pretrained(inference_model, peft_model_id)

# Reset the metric
metric = evaluate.load("accuracy")

inference_model.to(device)
inference_model.eval()
for step, batch in enumerate(tqdm(eval_dataloader)):
    batch = {k: v.to(device) for k, v in batch.items()}
    with torch.no_grad():
        outputs = inference_model(**batch)
    predictions = outputs.logits.argmax(dim=-1)
    predictions, references = predictions, batch["labels"]
    metric.add_batch(
        predictions=predictions,
        references=references,
    )

eval_metric = metric.compute()
print(eval_metric)
```

I tried switching to `PeftModelForSequenceClassification` but that did not resolve the issue.

Again, this issue also occurs for me outside of Colab, on my local machine, and running python scripts that fully exit after each execution (edit: [here is a repro](https://github.com/jchook/peft-issue-2171))

This issue also occurs when I try to fix the random seeds for numpy, torch, etc.

### Expected behavior

I would expect relatively similar test accuracy each time I run the test routine.

**Comment by JINO-ROHIT** [comment]
Hi @jchook , i think i know what the issue is, I debugged your code, seems like the reuse of inference_model variable twice is causing the issue, can you try this? it works for me, lmk.


```
import torch
from peft import PeftModelForSequenceClassification, PeftConfig
from transformers import AutoModelForCausalLM, AutoTokenizer

peft_model_id = "./output/lora_epoch_1"
config = PeftConfig.from_pretrained(peft_model_id)
base_model = AutoModelForSequenceClassification.from_pretrained(config.base_model_name_or_path)
tokenizer = AutoTokenizer.from_pretrained(config.base_model_name_or_path)

# Load the Lora model
inference_model = PeftModelForSequenceClassification.from_pretrained(base_model, peft_model_id)

# Reset the metric
metric = evaluate.load("accuracy")

inference_model.to(device)
inference_model.eval()
for step, batch in enumerate(tqdm(eval_dataloader)):
    batch = {k: v.to(device) for k, v in batch.items()}
    with torch.no_grad():
        outputs = inference_model(**batch)
    predictions = outputs.logits.argmax(dim=-1)
    predictions, references = predictions, batch["labels"]
    metric.add_batch(
        predictions=predictions,
        references=references,
    )

eval_metric = metric.compute()
print(eval_metric)
```

**Comment by JINO-ROHIT** [comment]
for added clarity and explanation, basically since you reuse the variable, the peft model keeps getting overwritten and thats causing inconsistencies i think

**Comment by jchook** [comment]
Thanks for the assistance. 

I don't believe this fixes the issue.

Starting from a fresh Colab session and using your updated test loop, here is the output of my first 10 consecutive executions of the test cell:

```
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 0.0}
{'accuracy': 1.0}
{'accuracy': 0.041}
{'accuracy': 0.736}
{'accuracy': 0.0}
{'accuracy': 0.033}
{'accuracy': 0.0}
{'accuracy': 0.04}
```

I can reproduce this using normal .py scripts on my local machine where the python interpreter fully exits between executions, so I also don't believe it has to do with lingering Colab session state.

**Comment by JINO-ROHIT** [comment]
hmmm @jchook thats weird, can you show me how youre evaluations for n executions.

I get consistent results for each run, you can find the example here - https://colab.research.google.com/drive/1bwhOStf2PZ7CfYP7x9hrticJ_rmUWkdN?usp=sharing

**Comment by jchook** [comment]
@JINO-ROHIT I very much appreciate the investigation.

> can you show me how youre evaluations for n executions.

The key is ensure models are reloaded on each execution. You can manually re-execute the entire test cell, or modify your code to load the LoRA and base model from within your run loop.

<details>
  <summary>See code changes here</summary>

### Faster execution for the sake of time

First, to make the test code run faster, I reduced the test set to only 100 examples in the dataloader config:

```diff
- dataset['test'] = dataset['test'].select(range(1000))
+ dataset['test'] = dataset['test'].select(range(100))
```

### Ensure your run loop re-loads the LoRA from disk each time

The bug appears to have something to do with loading from disk and applying to the base model. Also, make sure you have a sufficient size for `n` as I sometimes see `{accuracy:1.0}` for the first ~5 executions.

```python
import torch
from peft import PeftModelForSequenceClassification, PeftConfig
from transformers import AutoModelForSequenceClassification, AutoTokenizer
import evaluate
from tqdm import tqdm

n = 10  # Number of times to run the evaluation
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

peft_model_id = "./output/lora_epoch_1"
results = []

# Loop to run the evaluation `n` times
for run in range(n):
    config = PeftConfig.from_pretrained(peft_model_id)
    base_model = AutoModelForSequenceClassification.from_pretrained(config.base_model_name_or_path)
    tokenizer = AutoTokenizer.from_pretrained(config.base_model_name_or_path)

    inference_model = PeftModelForSequenceClassification.from_pretrained(base_model, peft_model_id)

    inference_model.to(device)
    inference_model.eval()

    # Reset the metric for each run
    metric = evaluate.load("accuracy")
    print(f"Run {run + 1}/{n}")

    for step, batch in enumerate(tqdm(eval_dataloader)):
        batch = {k: v.to(device) for k, v in batch.items()}
        with torch.no_grad():
            outputs = inference_model(**batch)
        predictions = outputs.logits.argmax(dim=-1)
        references = batch["labels"]
        metric.add_batch(
            predictions=predictions,
            references=references,
        )

    eval_metric = metric.compute()
    print(f"Evaluation result for run {run + 1}: {eval_metric}")
    results.append(eval_metric)
  
print("Final evaluation results:")
for result in results:
  print(result)
```
</details>

After ensuring that the LoRA is re-loaded on each execution and setting `n=10`, here are my results:

```
Final evaluation results:
{'accuracy': 0.0}
{'accuracy': 0.86}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 0.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
```

I'm able to consistently reproduce this issue on Colab and my local environment. Here is a reproduction of the issue using normal python scripts, ensuring python fully exits between each invocation of `test.py`: https://github.com/jchook/peft-issue-2171

You can see [the first 100 `test.py` invocation results](https://github.com/jchook/peft-issue-2171/blob/main/results.txt).

**Comment by JINO-ROHIT** [comment]
ahhh i see , not sure why then, the other bits of the code seem mostly okay, did pushing to hub and load help? 

ill try and see if something else works, meanwhile we can wait for @BenjaminBossan

**Comment by jchook** [comment]
> did pushing to hub and load help?

Apparently saving/loading from the Hub does not fix the issue either. See [this repro on Colab](https://colab.research.google.com/drive/1LUzX6RQ4NXXItWberVTG_w4Ad2Nx4EoA?usp=sharing).

```
Final results:
{'accuracy': 0.0}
{'accuracy': 0.525}
{'accuracy': 0.0}
{'accuracy': 0.99}
{'accuracy': 0.0}
{'accuracy': 0.9}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 0.73}
{'accuracy': 1.0}
```

However, changing the base model from DeBERTa V3 to RoBERTa seems to [resolve the issue](https://colab.research.google.com/drive/1N0orjR_c_laXi4KeEi5iNn9Gbdz55DdE#scrollTo=4d55c87d). So the problem may be DeBERTa-specific somehow.

```
Final results:
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
{'accuracy': 1.0}
```

**Comment by BenjaminBossan** [comment]
Thanks for reporting this error and thanks @JINO-ROHIT for also investigating.

Could you try adding the `pooler` to the `modules_to_save`? Right now, only the `classifier` is saved. So in the `LoraConfig`, pass `modules_to_save=["classifier", "pooler"]`. I suspect that the issue is that the pooler is randomly initialized for the classification task and that's why you see erratic behavior.

**Comment by jchook** [comment]
Thanks @BenjaminBossan, that fixes the issue 🎉

Makes sense, as DeBERTa does not have a pretrained pooler like many other BERT models do.

Are there any next steps on, e.g., adding documentation for this somewhere or making PEFT automatically detect that DeBERTa has `requires_grad=True` on the pooler layers? Or is this just something that folks should be aware of when using models like DeBERTa?

**Comment by BenjaminBossan** [comment]
I'm glad that this resolved the issue.

> is this just something that folks should be aware of when using models like DeBERTa?

Well, theoretically, users should be aware of this because they should get the following warning when they load the model:

> Some weights of DebertaV2ForSequenceClassification were not initialized from the model checkpoint at microsoft/deberta-v3-base and are newly initialized: ['classifier.bias', 'classifier.weight', 'pooler.dense.bias', 'pooler.dense.weight']

However, even if they read this warning, they need to be aware of what `modules_to_save` really does and that `pooler` is _not_ automatically added (in contrast to `classifier`). So it is indeed an easy mistake to make.

As to potential solutions: If you know a good place in the docs (maybe somewhere that you looked for this issue), feel free to let me know or create a PR to add it. I could see adding a paragraph here for instance:

https://huggingface.co/docs/peft/developer_guides/troubleshooting#randomly-initialized-layers

However, I can already predict that many users won't read the docs and still run into the same trap.

> making PEFT automatically detect that DeBERTa has requires_grad=True on the pooler layers?

Hmm, I don't think that this can really work, just because a layer has `requires_grad=True` does not mean that it should be added to `modules_to_save`.

**Comment by JINO-ROHIT** [comment]
oh wow, this was indeed an easy mistake to make, glad i learned something new here. Thanks guys

**Comment by BenjaminBossan** [comment]
@jchook I created a PR (#2175) to add a bit of extra info to the troubleshooting docs. Maybe this will help in the future if people search for this PEFT issue with DeBERTa. Other than that, IMO this doc section already explains the problem quite well. LMK if you have a suggestion where else to document this.
